File: hash/best.js
let genres = ["classic", "pop", "classic", "classic", "pop"]
let plays = [500, 600, 150, 800, 2500]

let map = new Map();

for (let [index, value] of genres.entries()) {
    if (!map.get(value)) map.set(value, [plays[index] , new Map().set(index, plays[index])]);
    else map.set(value, [map.get(value)[0] + plays[index], map.get(value)[1].set(index, plays[index])]);
}
let sorted = Array.from(map);
sorted.sort((a, b) => b[1][0] - a[1][0])
console.log(sorted)
let first = [ ...sorted[0][1][1].entries() ].sort((a,b) => b[1] - a[1])
let second = [ ...sorted[1][1][1].entries() ].sort((a,b) => b[1] - a[1])

let result = [first[0][0], first[1][0], second[0][0], second[1][0]]

File: hash/Best.java
package hash;

import java.util.*;

import javax.print.event.PrintEvent;

class Best {

    static class Member implements Comparable<Member>{
        public Integer idx;
        public Integer playCount;

        Member(Integer idx, Integer playCount) {
            this.idx = idx;
            this.playCount = playCount;
        }

        @Override
        public int compareTo(Member target) {
            if (this.playCount < target.playCount)
                return 1;
            if (this.playCount > target.playCount)
                return -1;
            if (this.idx > target.idx)
                return 1;
            return 0;
        }
    }

    static String[] genres = { "classic", "pop", "classic", "classic", "pop" };
    static int[] plays = { 500, 600, 150, 800, 2500 };

    public static void main(String[] args) {
        Map<String, Integer> count = new HashMap<>();
        Map<String, PriorityQueue<Member>> member = new HashMap<>();
        for (int i = 0; i < genres.length; i++) {
            if (count.get(genres[i]) == null) {
                count.put(genres[i], plays[i]);

                PriorityQueue<Member> pq = new PriorityQueue<>();
                pq.offer(new Member(i, plays[i]));
                member.put(genres[i], pq);
            } else {
                count.put(genres[i], count.get(genres[i]) + plays[i]);
                member.get(genres[i]).offer(new Member(i, plays[i]));
            }
        }

        //맵 정렬
        List<Map.Entry<String, Integer>> entriesList = new LinkedList<>(count.entrySet());
        entriesList.sort(new Comparator<Map.Entry<String, Integer>>() {
            @Override
            public int compare(Map.Entry<String, Integer> o1, Map.Entry<String, Integer> o2){
                return o2.getValue() - o1.getValue();
            }
        });
        List<Integer> result = new ArrayList<>();
        for(Map.Entry<String,Integer> entry : entriesList){
            Queue<Member> queue = member.get(entry.getKey());
            result.add(queue.poll().idx);
            if (!queue.isEmpty())
                result.add(queue.poll().idx);
        }
        int[] ans = result.stream().mapToInt(i -> i).toArray();
    }

}